perf(db) reuse first call of schema_state in translation function

hishamhm opened the pull request:
The translation function for plugins:select_by_cache_key needs to poll
schema_state to decide when to stop using the "migrating" version of the
operation.

Kong checks the schema state once on startup. This commit adds a new
operation, `DB:last_schema_state` to retrieve the last value obtained
(`DB:schema_state` remains the same, re-querying the live database). The
`plugins:select_by_cache_key` translation function first checks
`last_schema_state` before performing the more expensive operation using
`schema_state`.

In the usual scenario where Kong is properly migrated before startup, the use
of `last_schema_state` will be sufficient for disabling the translation
function.

See #4164.

p0pr0ck5 commented:
Just a quick note, I can confirm that this does indeed reduce the latency envelope as [expected](https://github.com/Kong/kong/issues/4164#issuecomment-453225899). Nice :eyes:!

thibaultcha commented:
Let's add some tests for this new method as well.

hishamhm commented:
> Let's add some tests for this new method as well.

Sure! Apparently the old method (or the DB module at large) didn't have any explicit tests either, so I added the beginnings of a unit test suite for `kong/db/init.lua` with a test for `last_schema_state`.

I also split the commit into `feat()` and `perf()` sections. It is still a fit for `master` in spite of `feat()` because the feature is internal.

thibaultcha commented on spec/01-unit/01-db/07-db.lua:
```diff
@@ -0,0 +1,99 @@
+      local last_state_3 = db:last_schema_state()
+
+      -- the latest object created by schema_state is the one cached
+      assert(state_2 == last_state_3)
```
`assert.equal` is preferred since it produces a graceful busted failure, while this assertion produces an error. Failed and error-ed tests are counted differently in busted.

hishamhm replied:
No, this produces as graceful failure as well (I just checked by changing to `~=`. I wanted to be explicit I'm checking for pointer equality.

Tieske replied:
`equal` does pointer equality, the `same` assertion only compares contents.

In this case a failed assertion will just give an 'assertion failed' error, where the `equal` assertion would provide feedback of the actual contents of the variables.

hishamhm replied:
> feedback of the actual contents of the variables.

And that feedback is misleading, because this test is not about their contents. In the case of two tables that are different objects but structurally equivalent (which is the case I'm trying to catch), it results in a message like

```
spec/01-unit/01-db/07-db.lua:93: Expected objects to be equal.
Passed in:
(table) {
  [legacy_is_014] = false
  [needs_bootstrap] = false
  [new_migrations] = {
    [1] = {
      [migrations] = { ... more }
      [namespace] = 'kong.db.migrations.core'
      [subsystem] = 'core' } } }
Expected:
(table) {
  [legacy_is_014] = false
  [needs_bootstrap] = false
  [new_migrations] = {
    [1] = {
      [migrations] = { ... more }
      [namespace] = 'kong.db.migrations.core'
      [subsystem] = 'core' } } }

stack traceback:
	spec/01-unit/01-db/07-db.lua:93: in function <spec/01-unit/01-db/07-db.lua:59>
```

If one reads the message above, what will they assume? First, they will assume the some deep field that's hidden under `... more` is different. If the tables _are_ structurally different, then one might assume that making them strucuturally identical would fix it, because it's the structures being shown and not the raw pointers, which is not the case.

Compare that to getting an `assertion failed`, going straight to the offending line in the test file and seeing that tables are being compared with `==` (plus, these assertions have comments).

I knew about `assert.equal` and `assert.same`, my choice for `==` was deliberate.

But sure, I can change the comments into assertion messages. :)

Tieske replied:
All opinions here. Best solution would be to change the assertion message and including the table-id in the `equal` assertion in Luassert.

bungle commented on kong/db/init.lua:
```diff
@@ -411,8 +411,18 @@ do
+
   function DB:schema_state()
-    return MigrationsState.load(self)
+    last_schema_state = MigrationsState.load(self)
```
The `MigrationsState.load` may return more than one return value. Do we care about them? Perhaps better error reporting / logging?

hishamhm replied:
Excellent catch! Fixed!
